### PR TITLE
Support discrete variables in get_dependencies() via ProvenanceTensor

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2957 
pair coded with @ordabayevy 

This refactors `get_dependencies()` to use @ordabayevy's `ProvenanceTensor` instead of `torch.autograd.grad`. This results in a cubic -> linear speedup, and support for tracking dependencies through discrete latent variables and through `.detach()` (fixing a previously known failure mode).

## Tested
- [x] refactoring covered by existing tests
- [x] fixed a known failure due to `.detach()`
- [x] added a new discrete test

### CODE REVIEW

File: pyro/ops/provenance.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+
+
+class ProvenanceTensor(torch.Tensor):
+    """
+    Provenance tracking implementation in Pytorch.
```
Comment (eb8680) [inline pyro/ops/provenance.py:9]: Maybe add slightly more documentation here explaining what extra metadata is attached and a reference to 'Wingate, David, et al. "Nonstandard Interpretations of Probabilistic Programs for Efficient Inference." NIPS. 2011'?

**Review by eb8680 — COMMENTED** [review]
Looks reasonable. Would this ever interact with the JIT in the process of autoguide initialization? If so, is that exercised by any existing tests of e.g. `AutoStructured`?

Also, are there any noteworthy edge cases of `__torch_function__` that would produce incorrect results or cause an error? For example, does it work with most built-in `nn.Module`s (e.g. `nn.Linear`) in `torch.nn`? Maybe add a test case with the Bayesian regression example to make sure it plays nicely with `nn.Linear` and `PyroModule`?

File: pyro/ops/provenance.py (hunk 1 above)
Comment (fritzo) [reply pyro/ops/provenance.py:9]: done

**Comment by fritzo** [comment]
> Would this ever interact with the JIT in the process of autoguide initialization?

`get_dependencies()` is called during autoguide initialization in `AutoStructured`  `AutoGaussian` which are already heavily tested with the jit and with nn.Modules in tests/infer/test_autoguide.py and tests/infer/autoguide/test_gaussian.py

> Also, are there any noteworthy edge cases of __torch_function__ that would produce incorrect results or cause an error?

@ordabayevy said the `.__torch_function__()` code is young. I expect there will be edge cases, and we can address them as they arise, but there are no know failures.

> does it work with most built-in nn.Modules ... ?

The `__torch_function__` layer is below and independent of the `nn.Module` layer. This is already tested in test_autoguide.py

**Comment by fritzo** [comment]
Blocked by https://github.com/pypa/setuptools/issues/2849

**Review by eb8680 — APPROVED** [review]
Lgtm

**Comment by fritzo** [comment]
Thanks for reviewing @eb8680 !
